### ISSUE TITLE
Preventing Track Segment Collisions?

### ISSUE THREAD

**Issue by justinlevi** [wavesjs/waves-ui#25]
This is such a great project! Everything seems very well organized and documented. Really impressive work. 

I'm just finding this project and starting to dig in. Before I get too deep, I was wondering if you could help point me to any examples in order to prevent track segment collisions? In other words, I don't want the segments to overlap when you drag them along a track. 

In the boilerplate > `Advanced Usage`example `[example-boilerplat.html](https://github.com/wavesjs/waves-ui/blob/master/examples/example-boilerplate.html)`, I can see that we're using `setBehavior` with a new SegmentBehavior() object which seems to handle the mouse interaction. Where would you add code to check collisions? 

How might you implement segment reordering while dragging as well?

Thanks again for any help as I start to wrap my brain around the code base.

**Comment by b-ma** [comment]
Hi,

Thanks for the feedback. To prevent segments collisions, you actually need to implement your own behavior (the default behavior provided by the library doesn't prevent that, as you have noticed).

I would probably do something like that:

```js
class PreventCollisionSegmentBehavior extends ui.behavior.BaseBehavior {
  constructor(segmentData) {
    super();
    // segmentData is a reference to the data defining all your segments
    this.segmentData = segmentData; 
  }

  // then reimplement _move, _resizeLeft, _resizeRight to clip the values of your 
  // edited segment to prevent the collision
  // for example (I didn't test the code, but here is the idea):
  _move(renderingContext, shape, datum, dx, dy, target) {
    const layerHeight = renderingContext.height;
    // current values
    const x = renderingContext.timeToPixel(shape.x(datum));
    const y = renderingContext.valueToPixel(shape.y(datum));
    const height = renderingContext.valueToPixel(shape.height(datum));
    // target values
    let targetX = Math.max(x + dx, 0);
    let targetY = y - dy;

    // lock in layer's y axis
    if (targetY < 0) {
      targetY = 0;
    } else if (targetY + height > layerHeight) {
      targetY = layerHeight - height;
    }

    // test against other segments
    this.segmentsData.forEach(neighbor => {
     if (datum !== neighbor) {
       // test left boundary
       if (dx < 0) { // segment is moved to the right
         const neighborEndTime = shape.x(neighbor) + shape.width(neigbor);
         const neighborEndPixel = renderingContext.timeToPixel(neighborEndTime);
         // clip segment position to right boundary of the neighbor 
         targetX = Math.max(targetX, neighborEndPixel)
       } else if (dx > 0) {
         const segmentWidth = renderingContext.timeToPixel(shape.width(datum));
         const segmentEndPixel = targetX + segmentWidth;
         const neightborStartPixel = renderingContext.timeToPixel(neighbor.x(datum));
         // currentSegment (x + width) must be < neighbor x
         if (neightborStartPixel < segmentEndPixel)
           targetX = neightborStartPixel - segmentWidth;
       }
     }
    });
 
    shape.x(datum, renderingContext.timeToPixel.invert(targetX));
    shape.y(datum, renderingContext.valueToPixel.invert(targetY));
  }
}
```

Then you use this behavior on your layer:
```
segmentLayer.setBehavior(new PreventCollisionSegmentBehavior());
```
The same operation should apply for the resize methods.

As for the reordering, I don't really get what you want to achieve, but probably it would also be a matter of making the behavior a bit more complex than the one provided in the library.

Let me know if it works for you.

**Comment by justinlevi** [comment]
This is fantastic! Thanks for the info. I'll give this a shot and let you know how it works out. 
Cheers!

**Comment by justinlevi** [comment]
@b-ma - I feel like a bit of an idiot so apologies up front. 
1. Cloning the repo outright, running `npm install` (yarn install) and then `npm run build` threw a bunch of errors. I tried to fix that by adding everything as a dev dependency and updating `babel` to `babel-core` in the bin/scripts. 

2. For some reason, even though I'm requiring the the `CollisionSegmentBehavior` in src/waves-ui.js, the resulting min and umd file doesn't seem to include the class? 

I have put together a PR here. 
https://github.com/wavesjs/waves-ui/pull/27

I'm probably just doing something super lame so thanks in advance. 

Also, any tips on how you might use waves-ui in a React project?

**Comment by justinlevi** [comment]
Looking at the Develop branch this morning I’m realizing many of the Babel updates have already been incorporated. I’ll give that a shot next

**Comment by b-ma** [comment]
Hi,

Yes, they is a bunch of changes in the `develop` branch that should be merge in the master soon.

As for your problem, if you properly exposed your module in `src/waves-ui.js`, and it looks like you had, it should work. 
Do you import the library as an es6 module using babel and browserify or some kind of build tool ? 
Did you link the library (`npm link`) to use your local version in the `node_modules` ?
 
These questions aside, I'm not sure if it's a good idea to put your code inside the library, because `behaviors` are generally very specific to an application and the ones provided in the library are more thought as very generic examples. But I would be very interested in finding a clean way to create some plugins as separate packages (that could be linked in the documentation) — because I'm a bit concerned with maintainability if the library starts merge every single possible behavior.

For example, using the plugin containing your behavior might looks like:
```
import * as ui from 'waves-ui';
import CollisionSegmentBehavior from 'waves-ui-collision-segment-behavior';

// ...
segmentLayer.setBehavior(new CollisionSegmentBehavior());
```

As for your question about React, I did that a few years ago so I guess its still possible, but React probably changed quite a bit since this time and I'm not a regular user of this library so... But I would say that as long as you can access an underlying DOM element it should be feasible. Let me know if you find a way on this, it would be interesting to document as maybe other people might be interested.

**Comment by justinlevi** [comment]
@b-ma Sorry, I should have been more clear. I was just running some tests and wanted to minimize the opportunity for me to mess something else up. So, I decided to just add the CollisionSegmentBehavior next to the existing behaviors. Wasn't really thinking this would actually get merged in. Sorry for the confusion there. 

I am really hoping to dig into this code in depth and help where I can so I really wanted to be able to get the development environment up and running. I was able to clone the develop branch and make a couple of small changes in in this PR https://github.com/wavesjs/waves-ui/pull/28 and then get the bundle working. 

I've done a few react projects with d3 so I imagine it will be similar with how I need to create a reference to a DOM node. I'll add a tutorial here once I get things up and running. 

Thanks again for the direction. Really appreciate the help. This library is fantastic!
